### ISSUE TITLE
[QA] Medical Device Re-ordering Tool Test Plan

### ISSUE THREAD

**Issue by tlei123** [department-of-veterans-affairs/va.gov-team#5952]
## Product/Feature Summary

[Provide brief summary of product/feature to be tested on Staging by VSA-QA. Also, be sure to include links to the Product/Feature Outline, Epic, and Pre-Launch Checklist.]

Product/Feature Outline: [GitHub doc](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/medical-device-tool/product-outline.md#how-to-access-and-test)
Product/Feature Epic: #3766
Product/Feature Launch Checklist: [GitHub doc](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/medical-device-tool/launch-checklist.md)
UX Design wireframes/comps: [Invision prototypes](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/403219934_Default_-_Desktop)

## SMEs

PM: Leah Keeler (@leahkeeler)
Design/Research: Riley Orr (@rileyorr) , Becca Walsh (@rebeccawalsh) 
Front-end: Amen Ra (@mojaray2k), Mahariel Rosario (@mr0sari0)
Back-end: Cameron Testerman (@camerontesterman)

## Proposed Scenarios & Staging Test-Accounts

- **Scenario 1** -- `vets.gov.user+1@gmail.com` (Greg):
  - [Both Permanent & Temporary Addresses pre-populated](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/408098735)
  - Both Batteries & Accessories pre-populated
- **Scenario 2** -- `vets.gove.user+228@gmail.com` (Mark): [No Temporary Address pre-populated](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/407910005)
- **Scenario 3** -- `vets.gov.user+38@gmail.com` (Pauline): [All items ineligible](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/409233432) \[Intro-page dead end]
- **Scenario 4** -- `vets.gov.user+14@gmail.com` (Jerry): [All Batteries ineligible](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/418131332) including both alert types:
    - can't reorder until
    - can't reorder rechargeable batteries
- **Scenario 5** -- `vets.gov.user+17@gmail.com` (Eddie): [All Accessories options ineligible](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/407889090) \[Accessories gate-page skipped], triggering the alert-type:
    - [can't order accessories until](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/407509410)
- **Scenario 6** -- `vets.gov.user+21@gmail.com` (Chad): [Vet deceased](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/409068827) \[Intro-page dead-end]
- **Scenario 7** -- `vets.gov.user+27@gmail.com` (William): [Vet not in DLC database](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/409462681) \[Intro-page dead-end]
- **Scenario 8** -- `vets.gov.user+45@gmail.com` (Sam): [No Accessories on record](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/411094793)

[Credentials](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/blob/master/Administrative/vagov-users/mvi-staging-users.csv) \[private-repo doc]

## Review

[Describe notes and comments derived from the review of QA activities with developers and product manager.]

### Start URL(s):

* Landing (static) page: \[provide link]
* Form Intro page: https://staging.va.gov/hearing-aid-batteries-and-accessories/introduction

## Definition of Done

- [x] Test Plan reviewed by Product Team.
- [x] Test Plan finalized per review feedback from above.
- [ ] QA-request ticket\* linked (if any) \[ticket-template coming soon]

**Comment by leahkeeler** [comment]
Tze, Here is the URL for process flow https://app.mural.co/t/vsa8243/m/vsa8243/1578519389467/11a5327c66e1fb78cac3871889239f23e1f62998and design https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens however, the update address step is being modified based on new requirement from DLC - they maintain 2 addresses, a permanent one and temporary one (snowbird use case). The final design for how to display/update will be nailed down this week. Thanks!

**Comment by leahkeeler** [comment]
Hi Tze, The designs are complete reflecting the two address scenario. Please let me know next steps. or if I missed anything.

Thanks!

**Comment by tlei123** [comment]
@leahkeeler, thanks much for the updates!

While I'm building the initial "Main Flow" test-case, please start filling-out the Staging Test-Accounts & User Scenarios/Flows above in the main Description.

The Staging test-accounts can be ”TBD” for any Scenarios for which Staging API test-data are not ready yet, but I'd like to see a complete list of Scenarios so that I can develop matching test-cases.

Feel free to ping me on Slack as desired/needed to discuss further.

Thanks much!

**Comment by camerontesterman** [comment]
Tze I'm going to forward a list of users you can use when building out your test cases. Few things to know: testing anything in staging right now that depends on an API call to the DLC will not work because our API has not been integrated with theirs yet. Testing locally can be achieved, but right now only works for a specific user, I can let know which user that is in the email. When we are integrated with their staging environment, I will let you know. This list of users I am going to send you is the same one I sent them, so these users will have medical devices and supplies available to order that will be able to be seen within the app, once integration with the DLC is complete.

You will have to enable mocking in order to get the service calls to function properly in the local environment, if you have any trouble with this just let me know and we can hop on a call and I can show you how to set it up.

**Comment by tlei123** [comment]
Thx, @camerontesterman !  I'm just circling back to this...

Quick question about those users you sent me:  Would you happen to know the user-scenarios they map to?  We have Greg (test-account 1) for "Intro, Veteran Information, and Confirm Address pages", and I'm hoping you could map the others.

If you don't know, that's fine; I'll circle back to @rebeccawalsh Becca tomorrow.

Thanks much!

**Comment by tlei123** [comment]
@rebeccawalsh, would you mind scheduling up a Zoom meeting\* w/ @camerontesterman & me to discuss the user-scenarios & test-accounts further please?

Based on review of the Invision screens, I'm seeing more user-scenarios/-flows than the 5 test-accounts provided by @camerontesterman:

1. Vet has both permanent & temporary address pre-populated (in Staging API test-data).
1. Vet has only permanent address pre-populated.
1. Vet has both batteries & accessories options pre-populated.
1. Vet has only batteries option(s) pre-populated. (Does Accessories Yes/No gate-page still show?)
1. Vet has only accessories option(s) pre-populated. (Does Batteries Yes/No get-page still display?)
1. Vet gets a batteries-not-eligible-until alert.
1. Vet gets an accessories-not-eligible-until alert.
1. Vet gets all-items-ineligible alert.
1. Vet gets deceased alert.
1. Vet gets not-on-file alert.

Some of the above scenarios can be combined (e.g., the batteries-/accessories-ineligible-until alerts) into other scenarios to minimize the number of separate test-accounts to set up test-data behind Staging API(s), and we can suss that out when we meet.

\*I think a 30-min. mtg. should suffice, but feel free to make it 1 hr if you/Cameron also have questions for me.  My Google Calender is free/busy public.

**Comment by tlei123** [comment]
@camerontesterman @rebeccawalsh @rileyorr @mojaray2k @mr0sari0 , please review the updated **Proposed Scenarios...** section in the **main Description** at the top \[_not_ the previous Comment above]...

Either check-off below, or add a Comment with questions/issues: 

- [ ] Rebecca
- [ ] Riley
- [ ] Amen
- [ ] Mahariel
- [ ] Cameron

I'll check-in on your team channel about these in a day or two.

Thanks much!

cc @leahkeeler

**Comment by tlei123** [comment]
@rileyorr @rebeccawalsh , I have some questions about the two design screens below:

1. [Devices aren't eligible](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/406260563) -- How does this alert differ from [None of your devices are eligible](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/406054376)?  They seem to belong to the same user/data scenario.
2. [Accessories aren't available](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/411094793) -- Can you confirm that we do need a additional user/data scenario for this?  It does seem like a different situation from [None of your hearing aid accessories are eligible](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/407889090).  \[See the current list \[still under team review] in the main Description under Proposed Scenarios section.]

Please let me know.  Thanks much!

cc @camerontesterman

**Comment by rebeccawalsh** [comment]
@tlei123 

1. They are the same - I was making mass changes yesterday and missed updating that info component on one of the pages. One page is simply an expanded view of the additional info components. I updated the component and it should be more clear now. 

2. One case is reflecting if they have accessories tied to their profile but they are unavailable to be ordered while the other reflects a Veteran who has no accessories tied to their profile at all. They are slightly different cases.

**Comment by tlei123** [comment]
@leahkeeler, please have team review the updated Proposed Scenarios in the main Description at the top...

Thanks to @rebeccawalsh's clarification above \[see previous 2 comments], we now have an 8th scenario for Accessories unavailable, which is different from Accessories ineligible.

I have preliminary test-cases "mocked-up" in TestRail for the 7 originally listed proposed scenarios \[which still need your team's review], but since it still looks like early days (what with potential further design changes, etc.), I'll wait for review/feedback/approval before continuing in TestRail.

Thanks much!

**Comment by camerontesterman** [comment]
Users have been added for the scenarios. They are listed [here](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/blob/master/teams/vsa/teams/benefits-memorials-2/qa-test-plan-users.md)

[Mock Data PR](https://github.com/department-of-veterans-affairs/vets-api-mockdata/pull/123)

[Deceased Vet PR](https://github.com/department-of-veterans-affairs/vets-api/pull/4097)

Both of the PRs above must be merged for the users to function as intended.

**Comment by tlei123** [comment]
Scenarios updated with test-accounts in Description, thanks to Cameron's new user-list.

**Comment by tlei123** [comment]
[Test Cases](https://dsvavsp.testrail.io/index.php?/suites/view/8&group_by=cases:section_id&group_order=asc&group_id=57) now developed for current form-flow and design-screens, but not sure if references are really final...
\[Any BaMs team-member who'd like to access these Test Cases should request VSA-BaM2 team-account credentials from @leahkeeler ]

Will check-in with Leah tmrw (Thu) at Shared Support Standup re. design/development statuses, and see about upcoming test-cases refinements.

P.S.  Deleted comments with outdated scenarios list & outdated main-flow test-case review checklist.

**Comment by tlei123** [comment]
Per last Thursday's Shared Support Standup mtg, @leahkeeler will alert me about any upcoming changes that might impact the current Test Cases.

**Comment by tlei123** [comment]
@leahkeeler, pls. review my updated Main Flow test-case, for editing both Permanent and Temporary addresses...

Once you're at the [Test Cases page](https://dsvavsp.testrail.io/index.php?/suites/view/8&group_by=cases:section_id&group_order=asc) on TestRail, click on the **Main Flow (C303)** should already be highlighted by default \[it's at the top of the cases-list].  Pls. check **Steps 5-11** on the right-panel for editing Perm & Temp addresses.

If you're satisfied, pls. tick-off the top checkbox under Definition of Done in the main Description up top.  Otherwise, pls. add a Comment with your feedback for further changes.

Thx much!

**Comment by leahkeeler** [comment]
Thanks Tze - looks good

**Comment by tlei123** [comment]
For Scenarios 1 & 2, I'm running into test-account/test-data problems:

- Scenario 1 (Main-Flow test-case) - `vets.gov.user+1@gmail.com` - gets "Can't order batteries" alert.  A Battery order had apparently been submitted, which blocks the battery-ordering step.  Also, although 3 Accessories are available for ordering, I'm realizing this would only support 3 cross-browser/-device test-runs for this test-case.
- Scenario 2 (Add Temporary Address) - `vets.gov.user+228@gmail.com` - similarly, I'm realizing one test-account would only support one browser-/device-specific test-run for this test-case.

I'm also running into Staging auth redirect problems that are blocking me from running the rest of the test-cases right now.

I'll retry tomorrow and see if the auth-redirect issues persist.  The rest of the scenarios should be testable once I can log in.

**Comment by leahkeeler** [comment]
Hi Tze, I am not sure what causes the redirects-? but I am wondering if the two issues you are running across can be solved by resetting the data for test users between each test run. I have added @camerontesterman and @bradl3yC to this ticket/ Thanks, Leah

**Comment by bradl3yC** [comment]
@mr0sari0 with the most recent API changes, can we remove the hard coded always show alert box on the battery reorder?

**Comment by tlei123** [comment]
@leahkeeler @camerontesterman , yeah, let's discuss these things further in our meeting tomorrow.

I pressed ahead with the subsequent test-case steps of my Main Flow runs, just to make some progress.  Scenario 2 runs all Passed.

I'm holding off on the Copy/Layout cases for now, 'cuz I saw some mismatches between Staging & the Design prototypes.  Right now, I'm surmising any copy/layout defects I log right away could turn out false-positive.  I'll find some time this week to double-check w/ Riley/Becca.

Talk to you tomorrow at our meeting.

**Comment by tlei123** [comment]
Functional test-runs are all complete.

Re. copy/layout test-runs, I'm running into login issues on Staging, and pinged @camerontesterman on Slack re. my [relevant convo](https://dsva.slack.com/archives/C5HP4GN3F/p1589394059424000) with folks in #vfs-frontend -- something about "test-accounts with multiple conflicting MHV accounts."  \[All Greek to me.].  I suggested he be the one to contact ESD about this.

**Comment by tlei123** [comment]
@camerontesterman , I could log into Staging today [no need to call ESD], and re. the specific-alerts Scenarios...

It does look like some Staging **mock-data fixes are needed**:
- **Scenario 3** -- `vets.gov.user+38@gmail.com` (Pauline): [All items ineligible](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/409233432) \[Intro-page dead end]
- **Scenario 5** -- `vets.gov.user+17@gmail.com` (Eddie): [All Accessories options ineligible](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/407889090) \[Accessories gate-page skipped], triggering the alert-type:
    - [can't order accessories until](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/407509410)
- **Scenario 6** -- `vets.gov.user+21@gmail.com` (Chad): [Vet deceased](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/409068827) \[Intro-page dead-end]
- **Scenario 7** -- `vets.gov.user+27@gmail.com` (William): [Vet not in DLC database](https://vsateams.invisionapp.com/share/6MVTG94WNH5#/screens/409462681) \[Intro-page dead-end]

Feel free to ping me on Slack for further discussion, otherwise just update with a new Comment when Staging mock-data are fixed.

Thx much!

**Comment by leahkeeler** [comment]
@tzelei You are correct - these are known issues that are pending a platform fix in forms systems for error messaging. This should be completed in next 1-2 days. - Leah

**Comment by tlei123** [comment]
@leahkeeler , copy/layout test-runs are completed, and reports have been generated for our upcoming Staging Review.

TestRail links:
- [Test Plan](https://dsvavsp.testrail.io/index.php?/plans/view/35)
- [Coverage for References](https://dsvavsp.testrail.io/index.php?/reports/view/87)\*
- [Summary (Defects)](https://dsvavsp.testrail.io/index.php?/reports/view/91)

Feel free to ping me on Slack \[or add Comment here] with any questions/problems.

\*Re. Coverage for References, pls. let me know if you can find more specific tickets for each of the Test Cases.  I added the Epic ticket to them, but VSP would like each Test Case to reference a separate ticket (if possible).  Thx much!

cc @pjhill

**Comment by tlei123** [comment]
@leahkeeler @pjhill , just re-tested Main-Flow case for bugfix #9093 / #9712 \[2 tics, same bug]:

TestRails links:
- [Test Plan Rd.2](https://dsvavsp.testrail.io/index.php?/plans/view/434)
- [Coverage for References Rd.2](https://dsvavsp.testrail.io/index.php?/reports/view/98) \[focus on Main Flow case w/ new refs]
- [Summary (Defects) Rd.2](https://dsvavsp.testrail.io/index.php?/reports/view/97) \[beautifully blank -- all runs Passed]

P.S.  I'll Close #9712 with a Comment linking to this update here.

**Comment by leahkeeler** [comment]
Hey @tlei123  - Can I close this ticket out? I think everything is passing right

**Comment by tlei123** [comment]
Yep, closing now.
